src/Npgsql/TypeHandlers/ArrayHandler.cs

```diff
@@ -209,130 +148,177 @@ public override ValueTask<Array> Read(NpgsqlReadBuffer buf, int len, bool async,
+            lengthCache = elemLengthCache;
+            lengthCache.Lengths[pos] = len;
+            return len;
         }
 
-        /// <summary>
-        /// Take care of multi-dimensional arrays and non-generic IList, we have no choice but to box/unbox
-        /// </summary>
-        int ValidateAndGetLengthNonGeneric(IList value, ref NpgsqlLengthCache lengthCache)
+        // Take care of multi-dimensional arrays and non-generic IList, we have no choice but to box/unbox
+        int ValidateAndGetLengthNonGeneric(ICollection value, ref NpgsqlLengthCache lengthCache)
         {
             var asMultidimensional = value as Array;
             var dimensions = asMultidimensional?.Rank ?? 1;
 
             // Leave empty slot for the entire array length, and go ahead an populate the element slots
             var pos = lengthCache.Position;
             lengthCache.Set(0);
-            var lengthCache2 = lengthCache;
+            var elemLengthCache = lengthCache;
             var len =
-                4 +       // dimensions
-                4 +       // has_nulls (unused)
-                4 +       // type OID
-                dimensions * 8 +  // number of dimensions * (length + lower bound)
-                value.Cast<object>().Sum(element => 4 + GetSingleElementObjectLength(element, ref lengthCache2));
-            lengthCache = lengthCache2;
+                4 +              // dimensions
+                4 +              // has_nulls (unused)
+                4 +              // type OID
+                dimensions * 8 + // number of dimensions * (length + lower bound)
+                4 * value.Count; // sum of element lengths
+            foreach (var element in value)
+                if (element != null && element != DBNull.Value)
+                    try
+                    {
+                        len += ElementHandler.ValidateObjectAndGetLength(element, ref lengthCache, null);
+                    }
+                    catch (Exception e)
+                    {
+                        throw MixedTypesOrJaggedArray(e);
+                    }
+            lengthCache = elemLengthCache;
             lengthCache.Lengths[pos] = len;
             return len;
         }
 
-        int GetSingleElementObjectLength([CanBeNull] object element, ref NpgsqlLengthCache lengthCache)
-        {
-            if (element == null || element is DBNull)
-                return 0;
-            try
-            {
-                return ElementHandler.ValidateObjectAndGetLength(element, ref lengthCache, null);
-            }
-            catch (Exception e)
-            {
-                throw new Exception("While trying to write an array, one of its elements failed validation. You may be trying to mix types in a non-generic IList, or to write a jagged array.", e);
-            }
-        }
+        public override Task Write(Array value, NpgsqlWriteBuffer buf, NpgsqlLengthCache lengthCache, NpgsqlParameter parameter, bool async)
```

Comment:
As with ValidateAndGetLength - does this ever get called, given that we override `WriteWithLength<TAny>` below? If not throw `NotSupportedException`?

Reply:
See my comment above.